### PR TITLE
Fix the non cython build for cpp extensions

### LINKED ISSUE #19668: Pip build test CI failures
I've seen a couple of these. Is this intermittent? Any idea what's going on?

https://travis-ci.org/pandas-dev/pandas/jobs/340640051

```
    running build_ext
    pandas._libs.algos: -> [['pandas/_libs/algos.c']]
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-nv04tz46-build/setup.py", line 741, in <module>
        **setuptools_kwargs)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/site-packages/setuptools/__init__.py", line 129, in setup
        return distutils.core.setup(**attrs)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/command/install.py", line 545, in run
        self.run_command('build')
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/travis/miniconda3/envs/pandas/lib/python3.6/distutils/command/build_ext.py", line 339, in run
        self.build_extensions()
      File "/tmp/pip-nv04tz46-build/setup.py", line 367, in build_extensions
        self.check_cython_extensions(self.extensions)
      File "/tmp/pip-nv04tz46-build/setup.py", line 364, in check_cython_extensions
        """.format(src=src))
    Exception: Cython-generated file 'pandas/_libs/algos.c' not found.
                    Cython is required to compile pandas from a development branch.
                    Please install Cython or download a release package of pandas.
    
```

### PR DESCRIPTION
**Pull request by hexgnu**
- [ ] closes #19668
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

cc @TomAugspurger @jreback 

This _should_ fix the problem with building with a non Cython environment. The problem is that we assume extensions are C everywhere and I didn't realize that this was a test path.

Also setup.py might be in need of a rethink, there's a lot of special conditions that I think could be cleaned up with some better formatting.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -331,19 +330,24 @@ class CheckSDist(sdist_class):
...
-                           source=cfile))
-                assert os.path.isfile(cfile), msg
+            pyx_files = [(self._pyxfiles, 'c'), (self._cpp_pyxfiles, 'cpp')]
+
+            for pyxfiles, extension in pyx_files:
+                for pyxfile in pyxfiles:
+                    sourcefile = pyxfile[:-3] + extension
+                    msg = ("{extension}-source file '{source}' not found.\n"
+                           "Run 'setup.py cython' before sdist.".format(
+                               source=cfile, extension=extension))
```
Comment (hexgnu) [inline setup.py]: this is a mistake fixing now

**Comment by TomAugspurger** [comment]
@chris-b1 do you have any experience with Cython and cpp? Does this look OK to you?

**Comment by TomAugspurger** [comment]
The pip build stuff looks OK here: https://travis-ci.org/pandas-dev/pandas/jobs/341700234#L2422

File: setup.py [hunk 2]
```diff
@@ -618,7 +622,8 @@ def pxd(name):
     '_libs.window': {
         'pyxfile': '_libs/window',
         'pxdfiles': ['_libs/skiplist', '_libs/src/util'],
-        'language': 'c++'},
+        'language': 'c++',
+        'suffix': '.cpp'},
```
Comment (chris-b1) [inline setup.py]: rather than adding `'suffix'` any reason not to base the suffix on language='c++'?

**Comment by chris-b1** [comment]
One question, but this looks good to me.  @hexgnu, certainly welcome any refactoring of `setup.py`, @jbrockmendel has also been doing some work in that direction, see for instance #19484

File: setup.py (hunk 2 above)
Comment (hexgnu) [reply setup.py]: I guess I just saw it as more of a configuration piece that belonged in a dictionary over branches in the code.

I could do something like:

```python
    if suffix == '.pyx':
        source_suffix = '.pyx'
    elif data.get('language', 'c') == 'c++':
        source_suffix = '.cpp'
    else:
        source_suffix = '.c'
```

But that feels a bit verbose. Thoughts?

Comment (chris-b1) [reply setup.py]: Not a huge deal, but that feels cleaner too me.  Reading the config I would probably guess ``.suffix` controls the suffix of the output of cython, which it woudln't.

**Comment by TomAugspurger** [comment]
This seems OK. Any objections @jreback? It'd be good to get that build going on Travis again.

File: setup.py [hunk 3]
```diff
@@ -331,19 +330,24 @@ class CheckSDist(sdist_class):
...
 
+    _cpp_pyxfiles = ['pandas/_libs/window.pyx']
+
     def initialize_options(self):
         sdist_class.initialize_options(self)
 
     def run(self):
         if 'cython' in cmdclass:
             self.run_command('cython')
         else:
```
Comment (jreback) [inline setup.py]: can you add some comments on what is going on here

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok. do you want to try to fix the msgpack c++ stuff while you are at it? IOW fold it into the same framework

**Comment by hexgnu** [comment]
K I moved msgpack building into the same loop and added a comment.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. hopefully this fixes the pip build.

**Pull request merged**